File: spec/cpp_stl_11/prereq/my_custom_fx.cpp
#include "my_custom_fx.h"

my_custom_fx_t::my_custom_fx_t(int p_key, bool p_flag, std::string p_some_bytes) {
    key = p_flag ? p_key : -p_key;
}

std::string my_custom_fx_t::decode(std::string src) {
    int len = src.length();
    std::string dst(len, ' ');

    for (int i = 0; i < len; i++)
        dst[i] = src[i] + key;

    return dst;
}

File: spec/rust/src/my_custom_fx.rs
extern crate kaitai;
use kaitai::CustomDecoder;

pub struct MyCustomFx {
    key: i32,
}

impl MyCustomFx {
    pub fn new(p_key: u8, p_flag: bool, _p_some_bytes: &[u8]) -> Self {
        if p_flag {
            Self { key: p_key as i32 }
        } else {
            Self {
                key: -(p_key as i32),
            }
        }
    }
}

impl CustomDecoder for MyCustomFx {
    fn decode(&self, bytes: &[u8]) -> Result<Vec<u8>, String> {
        let mut res = bytes.to_vec();
        for i in res.iter_mut() {
            *i = (*i as i32 + self.key) as u8;
        }
        Ok(res)
    }
}
